nemo_skills/finetuning/data_preparation_utils/preprocessing.py

```diff
@@ -0,0 +1,236 @@
+
+                if not self.add_incorrect and not line_dict["is_correct"]:
+                    continue
+
+                # for backward compatibility
+                if "generation" not in line_dict and "generated_solution" in line_dict:
+                    line_dict["generation"] = line_dict.pop("generated_solution")
+
+                if line_dict["generation"] in seen_predictions[line_dict["question"]]:
+                    continue
+
+                seen_predictions[line_dict["question"]].add(line_dict["generation"])
+                line_dict['filename'] = file_handles[lidx].name
+                samples.append(line_dict)
+
+        for handle in file_handles:
+            handle.close()
+
+        return samples
+
+    def process(self):
+        samples = []
+        if self.prediction_jsonl_files:
+            samples.extend(self._read_raw_data())
+        if self.preprocessed_dataset_files:
+            samples.extend(self._read_preprocessed_data())
+        with open(self.output_manifest_file, "wt", encoding="utf-8") as fout:
+            for sample in samples:
+                fout.write(json.dumps(sample) + "\n")
+
+
+class GroupSamples(BaseProcessor):
+    def __init__(self, group_key='question', **kwargs):
+        super().__init__(**kwargs)
+        self.group_key = group_key
+
+    def process(self):
+        samples = defaultdict(list)
+        with open(self.input_manifest_file, "rt", encoding="utf-8") as fin:
+            for line in fin:
+                sample = json.loads(line)
+                samples[sample[self.group_key]].append(sample)
+
+        with open(self.output_manifest_file, "wt", encoding="utf-8") as fout:
+            for groupped_samples in samples.values():
+                fout.write(json.dumps(groupped_samples) + "\n")
+
+
+class ShuffleAndDownsampleData(BaseProcessor):
+    def __init__(
+        self,
+        random_seed: int,
+        do_shuffle: bool,
+        num_samples: Optional[int] = None,
+        sampling_method: str = 'random',
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.sampling_method = sampling_method
+        self.num_samples = num_samples
+        self.random_seed = random_seed
+        self.do_shuffle = do_shuffle
+
+    def process(self):
+        groupped_samples = []
+        with open(self.input_manifest_file, "rt", encoding="utf-8") as fin:
+            for line in fin:
+                samples = json.loads(line)
+                groupped_samples.append(samples)
+
+        random.seed(self.random_seed)
+        if self.sampling_method == "random":
+            output_instances = list(chain(*groupped_samples))
+            if self.do_shuffle:
+                random.shuffle(output_instances)
+            output_instances = output_instances[: self.num_samples]
+        elif self.sampling_method == "fair":
+            soln_counter = 0
+            output_instances = []
+            while self.num_samples is not None:
```

Comment:
why is this a "while"? Also this does not work if num_samples > data size, it hangs forever